File: Algorithms/Sorting/selectionSort.py
import os
import random
import time

ARRAY_SIZE = 10
ELEM_LIMIT = 50
TIME_DELAY = 0.005


def initArrayValues() -> None:
    array = []

    for _ in range(ELEM_LIMIT):
        array.append(random.randrange(0, ARRAY_SIZE))

    return array


def barGraph(array: list) -> list:
    maxElement = max(array)

    for i in range(maxElement):
        for j in range(len(array)):
            if array[j] >= maxElement - i:
                print("*", end="")
            else:
                print(" ", end="")

        print()


def isSorted(array: list) -> bool:
    for i in range(len(array) - 1):
        if array[i] > array[i + 1]:
            return False

    return True


def getMinValIndex(array: list, buffer: int) -> int:
    min = array[buffer]
    minIndex = buffer
    found = False

    for i in range(buffer + 1, len(array)):
        if min > array[i]:
            min = array[i]
            minIndex = i
            found = True

    return minIndex if found == True else -1


def selectionSort(array: list) -> list:
    barGraph(array)

    for i in range(len(array)):
        minValIndex = getMinValIndex(array, i)

        if minValIndex >= i and minValIndex != -1:
            array[i], array[minValIndex] = array[minValIndex], array[i]

            time.sleep(TIME_DELAY)
            os.system("cls")
            barGraph(array)

    return array


def main() -> None:
    array = initArrayValues()
    array = selectionSort(array)

    while True:
        if isSorted(array):
            print("DONE")
            break
        else:
            array = selectionSort(array)


if __name__ == "__main__":
    main()

File: Algorithms/Sorting/selectionSort.c
#include <stdio.h>
#include <conio.h>
#include <stdlib.h>
#include <time.h>

#define ARRAY_SIZE 50
#define ELEM_LIMIT 10

typedef unsigned int LoopVar;
typedef char BufferClear;

enum Bool {
   TRUE = 1,
   FALSE = 0
};

void delay(int);
void swapVals(int*, int*);
int getMinValIndex(int*, int);
void initArrayValues(int*);
void printArray(int*, int);
enum Bool isSorted(int*);
void barGraph(int*);
void selectionSort(int*);

BufferClear temp;
enum Bool shouldRepeat;

int main(void) {
   static int array[ARRAY_SIZE];

   system("cls");

   initArrayValues(array);

   selectionSort(array);

   printf("\nRepeat Program? (1 = YES): ");
   scanf("%d", &shouldRepeat);
   scanf("%c", &temp);

   (shouldRepeat == TRUE) ? main() : printf("\nEND");
}

void delay(int sec) {
   clock_t stTime = clock();

   while (clock() < stTime + sec) { }
}

void swapVals(int* val1, int* val2) {
   int temp;

   temp = *val1;
   *val1 = *val2;
   *val2 = temp;
}

int getMinValIndex(int* array, int buffer) {
   LoopVar i;
   int min;
   int minIndex;
   enum Bool found = FALSE;

   minIndex = buffer;
   min = array[buffer];

   for (i = 1 + buffer; i < ARRAY_SIZE; i++) {
      if (min > array[i]) {
         min = array[i];
         minIndex = i;
         found = TRUE;
      }
   }

   return (found == TRUE) ? minIndex : -1;
}

void initArrayValues(int* array) {
   int arraySize;
   LoopVar i;

   srand(time(NULL));

   for (i = 0; i < ARRAY_SIZE; i++) {
      array[i] = rand() % ELEM_LIMIT;
   }
}

void printArray(int* array, int size) {
   LoopVar i;

   for (i = 0; i < size; i++) {
      printf("%-4d", array[i]);
   }
}

enum Bool isSorted(int* array) {
   LoopVar i;

   for (i = 0; i < ARRAY_SIZE - 1; i++) {
      if (array[i] > array[i + 1]) {
         return FALSE;
      }
   }

   return TRUE;
}

void barGraph(int a[ARRAY_SIZE]) {
   int max;
   LoopVar i, j;

   max = a[0];

   for (i = 1; i < ARRAY_SIZE; i++) {
      if (a[i] > max) {
         max = a[i];
      }
   }

   for (i = 0; i < max; i++) {
      for (j = 0; j < ARRAY_SIZE; j++) {
         if (a[j] >= max - i) {
            printf("*");
         } else {
            printf(" ");
         }
      }
      printf("\n");
   }
}

void selectionSort(int* array) {
   LoopVar i;
   int minValIndex;

   // printArray(array);
   // printf("\n");
   barGraph(array);
   getch();

   for (i = 0; i < ARRAY_SIZE; i++) {
      minValIndex = getMinValIndex(array, i);

      if (minValIndex >= i && minValIndex != -1) {
         swapVals(&array[i], &array[minValIndex]);

         // printArray(array);
         // printf("\n");
         delay(20);
         system("cls");
         barGraph(array);

         if (isSorted(array) == TRUE) {
            return;
         }
      }
   }

}